---
title: "Identify X or Z Chromosomes from genotype data"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---

```{r setup, include = F} 
library(tidyverse)
library(plyr)
library(broom)
library(dplyr)
library(ggplot2)
```

Required: 

* .vcf 
* vcftools in path
* sex of individuals in your dataset (tab delimited)

Make list of chromosomes/scaffolds/contigs found in your .vcf

Take care not to have problematic characters such as '|' in your scaffold names

The following is the code to run the vcftools --het analysis in a shell script or can be submitted to a job scheduler with the proper header.

This will result in an output directory containing observed and expected homozygosities, number of loci, and inbreeding coefficient (F) for each individual for each scaffold separately. Following the vcftools output, this script will concatenate all outputs and append scaffold information.


```{r, engine = 'bash', eval = F}
vcfile=yourgenotypes.vcf
chrfile=yourchr.txt
outdir=Homozygosities
homozygositiesfile=allhomozygosities.tsv


if [ ! -f "$outdir" ]; then
    mkdir $outdir
fi

cat $vcfile | grep -v "^#" | awk '{print $1}' | sort -u >$chrfile

for x in `cat $chrfile`
do
vcftools --vcf $vcfile --het --chr $x --out $x 
done

for x in `cat $chrfile`
do
cat $outdir/$x.het | awk -F '\t' -v i=$x -v OFS='\t' '{print 'i', $0}' | grep -v "O(HOM)" >>$homozygositiesfile
done

sed -i '1iScaffold\tIndv\tO.hom)\tE.hom)\tN\tF' $homozygositiesfile
```

Below are R scripts to read in the single output file from above and a file you produce which contains the individual name (matching exactly with the individual name in the .vcf) and sex and appends sex information for all individuals to the augmented vcftools .het output. Additional columns are created to calculate the observed and expected heterozygosities, and observed and expected proportion heterozygosities. The last line selects only the relevant columns (Scaffold, Sex, and proportion of observed heterozygocities).

```{r results = 'hide'}
sexes <- read_tsv("Sex_Chr/Test/indv_sex.tsv", col_names=T)
heterozygosities <- read_tsv("Sex_Chr/Test/allhomozygosities.tsv", col_names=T) %>% 
  filter(N > 100) %>%
  inner_join(sexes) %>% 
  mutate(O.het =  N - O.hom) %>% 
  mutate(E.het = N - E.hom) %>%
  mutate(PE.het = E.het/N) %>% 
  mutate(PO.het = O.het/N) %>% 
  select(Scaffold, Sex, PO.het)
```

Below is a way to run a two-sample T test comparing the proportion of observed heterozygosities of females and males.

```{r results = 'hide'}
data_summary <- function(data, varname, groupnames){
  require(plyr)
  summary_func <- function(x, col){
    c(mean = mean(x[[col]], na.rm=TRUE),
      sd = sd(x[[col]], na.rm=TRUE))
  }
  data_sum<-ddply(data, groupnames, .fun=summary_func,
                  varname)
  data_sum <- rename(data_sum, c("mean" = varname))
 return(data_sum)
}

df <- data_summary(heterozygosities, varname = "PO.het", groupnames = c("Scaffold", "Sex")) %>% 
  pivot_wider(id_cols = Scaffold, names_from = Sex, values_from = c("PO.het", "sd")) %>% 
  drop_na()

scaffKeepers <- df$Scaffold

est <- heterozygosities %>% 
  filter(Scaffold %in% scaffKeepers) %>% 
  group_by(Scaffold) %>% 
  do(tidy(t.test(.$PO.het ~ .$Sex)))

df <- df %>% 
  inner_join(est, by = "Scaffold") %>% 
  mutate(Significant = ifelse(p.value < 0.001, "p-value < 0.001", ifelse(p.value >= 0.001, "p-value >= 0.001", "no")))
```

Create a scatter plot using the male proportion heterozygosity and the female heterozygosity with the standard deviations as lines on the x and y axes respectively. 

```{r}
p <- ggplot(df, aes(x = PO.het_M, y = PO.het_F, color=Significant)) + 
  geom_pointrange(aes(ymin = PO.het_F - sd_F, ymax = PO.het_F + sd_F)) +
  geom_errorbarh(aes(xmax = PO.het_M + sd_M, xmin = PO.het_M - sd_M, height = 0)) +
  labs(y="Heterozygosity (Female)", x = "Heterozygosity (Male)") + 
  scale_color_manual(values = c("#C93312", "#899DA4")) +
  theme_bw()

p + theme(legend.title = element_blank()) 
```

This is the same graph with the axes switched to be cute and put the male on the Y axis and the female on the X axis.

```{r}
p_switched <- ggplot(df, aes(x = PO.het_F, y = PO.het_M, color=Significant)) + 
  geom_pointrange(aes(ymin = PO.het_M - sd_M, ymax = PO.het_M + sd_M)) +
  geom_errorbarh(aes(xmax = PO.het_F + sd_F, xmin = PO.het_F - sd_F, height = 0)) +
  labs(y="Heterozygosity (Male)", x = "Heterozygosity (Female)") + 
  scale_color_manual(values = c("#C93312", "#899DA4")) +
  theme_bw()

p_switched + theme(legend.title = element_blank()) 
```

This table show the X or Z linked scaffolds based on these analyses.

```{r}
XZ_linked <- df %>% 
  filter(p.value <= 0.001) %>% 
  filter(PO.het_F < 0.05) %>% 
  select(PO.het_F, PO.het_M, sd_F, sd_M, p.value, method)

XZ_linked
```

